### PR TITLE
Refresh routing doco

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Connects to Particular/V6Launch#68

Resolves https://github.com/Particular/docs.particular.net/issues/1721

 * Redesign the main routing page
 * Remove the old message owner page and include the content in the main routing page
 * Review the main pub-sub page
 * Split MSMQ scale out doco into routing (file-based instance mapping) and scale-out proper
 * Add a doco on routing extensibility

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@weralabaj @bording @DavidBoike @timbussmann please review.

File: Snippets/Core/Core_5/ScaleOut/IndividualizeQueues.cs [hunk 1]
```diff
@@ -6,7 +6,7 @@ class IndividualizeQueues
     {
         void UniqueQueuePerEndpointInstance(BusConfiguration busConfiguration)
         {
-            #region UniqueQueuePerEndpointInstance 5.2
+            #region UniqueQueuePerEndpointInstance
```
Comment (weralabaj) [inline Snippets/Core/Core_5/ScaleOut/IndividualizeQueues.cs]: so this api is available from 5.0?

Comment (SzymonPobiega) [inline Snippets/Core/Core_5/ScaleOut/IndividualizeQueues.cs]: My mistake. I've committed too much. I'll revert this change.

File: nservicebus/messaging/publish-subscribe/index.md [hunk 2]
```diff
@@ -31,14 +31,14 @@ Or in simpler terms
 
 ## Mechanics
 
-Depending on the features provided by a given transport there are two possible implementations of Publish-Subscribe mechanics: "Persistence based" and "Native".
+Depending on the features provided by a given transport there are two possible implementations of Publish-Subscribe mechanics: "Emulated" and "Native".
```
Comment (weralabaj) [inline nservicebus/messaging/publish-subscribe/index.md]: > "Persistence based" => "Emulated"

I'm not sure how much people are used to those terms, but I don't see a good reason for changing it. If we insist, then I'd suggest a different name than "emulated" (like "NSB pub-sub" or "non-native").

File: nservicebus/messaging/publish-subscribe/index.md [hunk 3]
```diff
@@ -48,20 +48,17 @@ Available subscription persistences include
  * [InMemory](/nservicebus/persistence/in-memory.md)
  * [Azure Storage](/nservicebus/azure-storage-persistence)
 
-Transports that require persistences
-
- * [MSMQ](/nservicebus/msmq)
- * [SQLServer](/nservicebus/sqlserver)
- * [Azure Storage Queues](/nservicebus/azure-storage-queues)
-
+The emulated publish-subscribe is used by the unicast transports ({include:transports-unicast}).
```
Comment (weralabaj) [inline nservicebus/messaging/publish-subscribe/index.md]: Here we could say "NServiceBus provides/implements publish-subscribe for the following transports:

Comment (weralabaj) [inline nservicebus/messaging/publish-subscribe/index.md]: > unicast transports

I believe this is the first time "unicast" is used in this doco. Maybe it's not necessary? If necessary, we need to explain it first.

File: nservicebus/messaging/routing-extensibility.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+reviewed: 2016-08-17
+component: Core
+tags:
+- routing
+- extensibility
+related:
+- nservicebus/messaging/routing
+---
+
+partial:content
```
Comment (weralabaj) [inline nservicebus/messaging/routing-extensibility.md]: > The routing system can be extended

What for? Could we add an intro paragraph to explain when to use it?

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+snippet:RoutingExtensibility-TriggerEndpointShutdown 
+
+## Event routing
+
+As mentioned in the [general routing documentation](/nservicebus/messaging/routing.md), event routing differs depending on the transport capabilities. 
+
+Transports supporting the [Publish-Subscribe](/nservicebus/messaging/publish-subscribe/) pattern natively ({include: transports-multicast}) implement the event routing themselves. Refer to specific transport documentation for details on extensibility points.
+
+Transports without that support ({include: transports-unicast}) relay on NServiceBus core routing for event delivery. To emulate multicast routing on top of a unicast transport NServiceBus uses. The key concept is the collection of publishers. For each event it contains information on the logical endpoint that publishes it. Routing extensions can access the publishers collections from the `Feature` set up code:
```
Comment (weralabaj) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: > Transports without that support ({include: transports-unicast}) relay on NServiceBus

relay => rely

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+snippet:RoutingExtensibility-Publishers
+
+The source parameter has the same meaning and effect as in the routes collection.
+
+The publishers collection behaves the same way as the route table with regards to threading and atomicity.
+
+## Physical routing
+
```
Comment (weralabaj) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: Here also some intro paragraph would be useful - when and why would you use it?

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+Transports supporting the [Publish-Subscribe](/nservicebus/messaging/publish-subscribe/) pattern natively ({include: transports-multicast}) implement the event routing themselves. Refer to specific transport documentation for details on extensibility points.
+
+Transports without that support ({include: transports-unicast}) relay on NServiceBus core routing for event delivery. To emulate multicast routing on top of a unicast transport NServiceBus uses. The key concept is the collection of publishers. For each event it contains information on the logical endpoint that publishes it. Routing extensions can access the publishers collections from the `Feature` set up code:
+
+snippet:RoutingExtensibility-Publishers
+
+The source parameter has the same meaning and effect as in the routes collection.
+
+The publishers collection behaves the same way as the route table with regards to threading and atomicity.
```
Comment (weralabaj) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: > The publishers collection behaves the same way as the route table with regards to threading and atomicity.

Better as 

The publishers collection is thread-safe and all operations on that collection are atomic.

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md (hunk 6 above)
Comment (weralabaj) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: Also, how's that related to Command and Event routing? From the structure it seems like it's a 3rd type ;)

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+Routing extensions can influence the physical routing by modifying the endpoint instances collection:
+
+snippet:RoutingExtensibility-Instances
+
+The source parameter has the same meaning and effect as in the routes collection.
+
+The instances collection is thread-safe. It allows registering multiple instance of a given endpoint. In case there is more than one, message distribution is involved. 
+
+## Message distribution
```
Comment (weralabaj) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: How about calling it "Routing messages when scaling out"?

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+When there are multiple instances of a single logical endpoint registered in the routing system, each outgoing message has to undergo the distribution process to determine which instance is going to receive this particular message. All NServiceBus messages are always delivered to a single physical instance of any given logical receiver. By default a round-robin algorithm is used to determine the destination. Routing extensions can override this behavior by registering a custom `DistributionStrategy` for a given destination endpoint.
```
Comment (weralabaj) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: > When there are multiple instances of a single logical endpoint registered in the routing system, each outgoing message has to undergo the distribution process to determine which instance is going to receive this particular message. All NServiceBus messages are always delivered to a single physical instance of any given logical receiver. 

Every message is always delivered to a single physical instance of the logical endpoint. When the endpoint is scaled-out (i.e. there are multiple instances of a single logical endpoint), then NServiceBus by default determines the destination for a message in a round-robin fashion.

File: nservicebus/messaging/routing.md [hunk 10]
```diff
@@ -1,172 +1,68 @@
...
-snippet:Routing-DynamicRoutes
-
-Following example shows how to implement a shared-store based routing where destinations of messages are managed e.g. in a database.
-
-snippet:Routing-CustomRoutingStore
-
-NOTE: The function passed to the `SetFallbackRoute` call is executed **each time** a message is sent so it is essential to make sure it is performant (e.g. by caching the results if getting the result requires crossing a process boundary).
+## Routing concepts
 
+NServiceBus routing consists of two layers, logical and physical. Logical routing defines which logical endpoint should receive a given outgoing message. Physical routing defines to which actual physical instance of the selected endpoint should the message be delivered. While logical routing is a developer's concern, physical routing is controlled by operations. Apart from MSMQ transport physical routing usually is handled automatically by NServiceBus and does not require any configuration. To read more about physical routing with MSMQ go [here](/nservicebus/msmq/routing.md).
```
Comment (weralabaj) [inline nservicebus/messaging/routing.md]: Add one paragraph above explaining what is routing ;)

File: nservicebus/messaging/routing.md [hunk 11]
```diff
@@ -1,172 +1,68 @@
...
 
-### Endpoint instance mapping layer
 
-The purpose of the endpoint instance mapping layer is to provide information about physical deployments of a given endpoint. Endpoint instance mapping is optional for most transports in most cases because of built-in conventions. An example of situation where it is required is a scaled out endpoint using MSMQ transport. Other endpoints that send messages to it need to be aware of its instances.
+## Command routing
 
+As described [here](/nservicebus/messaging/messages-events-commands.md), NServiceBus distinguishes several types of messages. Command messages are always routed to a single destination.
```
Comment (weralabaj) [inline nservicebus/messaging/routing.md]: > Command messages are always routed to a single destination.

Single logical endpoint or something along those lines would be more consistent with other articles.

File: nservicebus/messaging/routing.md [hunk 12]
```diff
@@ -1,172 +1,68 @@
...
 
+The per-namespace routes override assembly-level routes and the per-type routes override both namespace and assembly routes.
 
-#### Static mapping
 
-The `EndpointInstances` class provides a method that allows the registration of a static mapping. This API is useful in specific scenarios where developers need to take control over a certain mapping.
+### Overriding the destination
 
-snippet:Routing-StaticEndpointMapping
+In specific cases the routing configuration can be overridden when sending a given message (e.g. send to a particular address or send to self). Refer to documentation on [sending](/nservicebus/messaging/send-a-message.md) for further details.
```
Comment (weralabaj) [inline nservicebus/messaging/routing.md]: What for?

File: nservicebus/messaging/routing.md [hunk 13]
```diff
@@ -1,172 +1,68 @@
...
 
 
-#### Dynamic mapping
+## Event routing
 
-Dynamic mapping is meant to provide a convenient extension for NServiceBus add-ons. To add a dynamic mapping rule pass a function that takes an endpoint name and returns a collection of endpoint instance names.
+Published event messages are routed to a single instance of each subscriber endpoint. However, before the message is published and delivered, the subscriber has to express its interest in a given type of event.
```
Comment (weralabaj) [inline nservicebus/messaging/routing.md]: > Published event messages are routed to a single instance of each subscriber endpoint

Events can be received by multiple logical endpoints, however in case of scale-out each event will be received only by one physical instance of any logical endpoint.

File: nservicebus/msmq/routing.md [hunk 14]
```diff
@@ -0,0 +1,25 @@
+---
+title: Physical routing with MSMQ
```
Comment (weralabaj) [inline nservicebus/msmq/routing.md]: How about scaling-out with MSMQ?

File: nservicebus/messaging/publish-subscribe/index.md (hunk 2 above)
Comment (SzymonPobiega) [inline nservicebus/messaging/publish-subscribe/index.md]: Changing to well-known "persistence-based message-driven"

File: nservicebus/azure-service-bus/multiple-namespaces-support.md [hunk 15]
```diff
@@ -52,7 +52,7 @@ Exactly two namespaces have to be configured when using `FailOverNamespacePartit
 
 ## Cross namespace routing
 
-NServiceBus allows to specify destination addresses using an `"endpoint@physicallocation"` in various places such as the [Send](/nservicebus/messaging/send-a-message.md) and [Routing](/nservicebus/messaging/routing.md) API or the [MessageEndpointMappings](/nservicebus/messaging/message-owner.md). In this notation the `physicallocation` section represents the location where the endpoint's infrastructure is hosted, such as a machine name or a servicebus namespace.
+NServiceBus allows to specify destination addresses using an `"endpoint@physicallocation"` in various places such as the [Send](/nservicebus/messaging/send-a-message.md) and [Routing](/nservicebus/messaging/routing.md). In this notation the `physicallocation` section represents the location where the endpoint's infrastructure is hosted, such as a machine name or a servicebus namespace.
```
Comment (timbussmann) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: I did not yet read the updated routing page, but since we explicitly disallow `RouteTo` parameters containing an "@" except for MEM, that would have to be an API declared by ASB similar to `.AtMachine` which should be documented here? Do we need to raise an issue on ASB about the requirements to support code first API for multi-namespace support (similar to the schema support issue @sqltransport)?

File: nservicebus/messaging/publish-subscribe/controlling-what-is-subscribed.md [hunk 16]
```diff
@@ -7,7 +7,7 @@ component: Core
 
 ## Automatic subscriptions
 
-The default mode for managing subscriptions is "auto subscribe". When a subscriber endpoint is started it determines to which events it needs to subscribe. It then sends subscription messages to the [owning endpoint](/nservicebus/messaging/message-owner.md) for those messages.
+The default mode for managing subscriptions is "auto subscribe". When a subscriber endpoint is started it determines to which events it needs to subscribe. It then sends subscription messages to the publisher endpoint for those messages.
```
Comment (timbussmann) [inline nservicebus/messaging/publish-subscribe/controlling-what-is-subscribed.md]: not related directly to this change, but should the list starting with line 16 also contain a point that message driven pub/sub also requires registered routes for autosubscribe to work?

File: nservicebus/messaging/publish-subscribe/index.md [hunk 17]
```diff
@@ -71,22 +68,14 @@ Subscriber2->Publisher: Subscribe to Message1
...
-
-![](mechanics-persistence-subscribe.svg)
+NOTE: In Versions 5 and below the *subscribe* message contains only the transport address of the subscriber. In Versions 6 and above the message contains also the information about the logical endpoint which is requesting the subscription. This information allows Version 6 and above endpoints to publish events to scaled out MSMQ endpoints without the need for a distributor in the middle. Unfortunately since Version 5 and below endpoints are not aware of this information, they require a distributor to be placed in front of a scaled out endpoint in order to maintain the correct semantics of event messages (being delivered to a single instance of an endpoint). To learn more see the [MSMQ scaling out page](/nservicebus/msmq/scalability-and-ha)
```
Comment (timbussmann) [inline nservicebus/messaging/publish-subscribe/index.md]: that sounds like something that should be in a msms specific page? Imho too much information/details for most users.

File: nservicebus/messaging/publish-subscribe/index.md [hunk 18]
```diff
@@ -168,6 +152,6 @@ NOTE: Version 2.X required a perfect match. This should make it easier to upgrad
 
 In some circumstances it may not be desirable to allow any endpoints to subscribe to a given publisher or event. NServiceBus provides a way to intervene in the subscription process and decide whether a given client should be allowed to subscribe to a given message.
 
-NOTE: Subscription authorization is only available when using transports that require persistence based publish-subscribe.
+NOTE: Subscription authorization is only available when using transports that don't support native publish-subscribe.
```
Comment (timbussmann) [inline nservicebus/messaging/publish-subscribe/index.md]: I don't like the negation in the sentence. Shouldn't that be something like

> Subscription authorization is not supported when using transports with native publish-subscribe.

or

> Subscription authorization is only available when using transports with persistence-based message-driven Publish Subscribe.

File: nservicebus/messaging/routing-extensibility_content_core_[,6).partial.md [hunk 19]
```diff
@@ -0,0 +1,13 @@
+The routing system can be extended by means of providing a routing configuration from source different than standard (`app.config` file) either by using a configuration source:
```
Comment (timbussmann) [inline nservicebus/messaging/routing-extensibility_content_core_[,6).partial.md]: should that be sources?

would "from different sources than" read better?

> (`app.config` file)

why not mentioning the code first API?

How about:

> The routing system can be extended by providing custom sources of routing information to extend or replace the standard routing configuration APIs.

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+The route table API is thread-safe and atomic, meaning either all the changes from the passed collection or successfully applied or none is.
+
+The routing system prevents route ambiguity. If the new or replaced routes conflict with existing ones, an exception is thrown. It is up to the route extension to deal with that exception but usually it is best practice to trigger the endpoint shutdown to prevent misrouting of messages.
+
+snippet:RoutingExtensibility-TriggerEndpointShutdown 
+
+## Event routing
+
+As mentioned in the [general routing documentation](/nservicebus/messaging/routing.md), event routing differs depending on the transport capabilities.
```
Comment (timbussmann) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: Not a huge fan of the "As mentioned in ..." for docs.

**Comment by timbussmann** [comment]
made some comments. Looks good in general to me. I especially like that you removed that diagram from the routing page :+1: . There are quite a few changes so it's quite hard to grasp all of them. I think I'll have to read the pages again afterwards to see how the "flow" and scope of the pages really work out but I like the direction this is going.

File: nservicebus/azure-service-bus/multiple-namespaces-support.md (hunk 15 above)
Comment (SzymonPobiega) [inline nservicebus/azure-service-bus/multiple-namespaces-support.md]: @timbussmann good catch. With the `@` validation the multi-namespace support in azure is not really usable (with code-first API)

File: nservicebus/messaging/publish-subscribe/controlling-what-is-subscribed.md (hunk 16 above)
Comment (SzymonPobiega) [inline nservicebus/messaging/publish-subscribe/controlling-what-is-subscribed.md]: Added that.

File: nservicebus/messaging/publish-subscribe/index.md (hunk 17 above)
Comment (SzymonPobiega) [inline nservicebus/messaging/publish-subscribe/index.md]: I've checked an this is already covered by MSMQ scale out page for V6. Removing.

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md (hunk 6 above)
Comment (SzymonPobiega) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: The structure of the doco is based on the structure of the API (which has four entry points). Not sure how to express that.

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md (hunk 8 above)
Comment (SzymonPobiega) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: It is not entirely same thing as when scaling out on brokers you don't need distribution. Distribution should be a well-known term, at least to people who are on MSMQ.

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md (hunk 9 above)
Comment (SzymonPobiega) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: I reworded that.

File: nservicebus/msmq/routing.md (hunk 14 above)
Comment (SzymonPobiega) [inline nservicebus/msmq/routing.md]: I think there is value in keeping them separate. It is perfectly valid to use the instance mapping file without even thinking about scaling out.

**Comment by SzymonPobiega** [comment]
@timbussmann @weralabaj I think I addressed most (if not all) of your comments.

**Comment by SzymonPobiega** [comment]
Unfortunately it can't be merged yet because it uses APIs which are not in the current beta package :(

**Comment by SzymonPobiega** [comment]
Updated to beta8

**Comment by SzymonPobiega** [comment]
I removed the inline includes. Should be good to merge now.

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md [hunk 21]
```diff
@@ -41,7 +41,7 @@ The publishers collection is thread-safe and all operations on that collection a
 
 Physical routing is responsible for mapping the destination logical endpoint to the actual transport address (queue name). 
 
-When using a broker transport ({include:transports-broker}), the physical routing is entirely managed by NServiceBus and does not require any configuration.
+When using a [broker transport](http://docs.particular.net/nservicebus/transports/#types-of-transports-broker-transports), the physical routing is entirely managed by NServiceBus and does not require any configuration.
```
Comment (SimonCropp) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: links should not be fully qualified

File: nservicebus/msmq/routing_content_core_[,6).partial.md [hunk 22]
```diff
@@ -0,0 +1,7 @@
+Host name of the machine running each NServiceBus endpoint can be specified directly in the message-endpoint mapping configuration section by adding `@machine` suffix.
+
+snippet:endpoint-mapping-msmq
+
+When using MSMQ, if there is no `@machine` part, NServiceBus assumes the configured endpoint runs on a local machine.
+
+MSMQ
```
Comment (SimonCropp) [inline nservicebus/msmq/routing_content_core_[,6).partial.md]: @SzymonPobiega  why is msmq here?
